### ISSUE TITLE
Potential instruction to help with u32 safety checking

### ISSUE THREAD

**Issue by bobbinth** [maticnetwork/miden#132]
For most `u32` operations the "safe" variants are much more expensive than the "unsafe" variations - frequently at least 5x or even 10x more expensive. This might make the safe variants much less desirable to use. One of the reason for this is that asserting that a value is a `u32` value is emulated by a sequence of 3 operations: `U32SPLIT EQZ ASSERT`. Thus, many instructions which work with 2 operands, asserting that the operands are 32 bit values requires 7 or 8 operations.

We could make safe operations much more efficient by introducing `U32ASSERT2` operation in the VM. This operation would check whether the top two items on the stack are `u32` values and if either one of them is not, return an error. Thus, we'd be able to do most safe operations at just 2x the cost of the unsafe ones.

From AIR constraint standpoint it shouldn't be too difficult to support this - i.e., this would require 4 16-bit range checks, which we need to support for other operations anyway (e.g. u32 mul).

**Comment by 0xKanekiKen** [comment]
If no one else is looking into it, then I'd like to pick this up?

**Comment by bobbinth** [comment]
That would be awesome! I do want to say that this might be a fairly involved task. So, maybe worth splitting it into a couple of PRs. For example, the first (simpler) PR could be to add `u32assert2` instruction to Miden assembly, and only after that add it to the VM's instruction set.

**Comment by 0xKanekiKen** [comment]
Actually, that's the plan! I will start working on adding ```u32assert2``` instruction to Miden assembly and submit a PR. Just wanted to know your thoughts on how are we planning to reduce the overall VM cycles. For ex: adding two u32 value unsafely will require 1 VM cycle and likewise doing it safely would costs us 7 VM cycles (1 for addition and 3 each for asserting u32). How can we do this in 2 VM cycles("_we'd be able to do must safe operations at just 2x the cost of the unsafe ones._")? Are we planning to do the operation unsafely first and then check of the output is u32 or not? Again this method will cost 4 VM cycles. Apologies if I'm missing something. Thanks.

**Comment by bobbinth** [comment]
> Actually, that's the plan! I will start working on adding `u32assert2` instruction to Miden assembly and submit a PR.

Great!

> For ex: adding two u32 value unsafely will require 1 VM cycle and likewise doing it safely would costs us 7 VM cycles

Once `u32assert2` is implemented as a native VM operation (i.e., an operation which takes 1 cycle) the cost of safe operations will be reduced significantly. But you are right, for operations where we need to check that output is in the right range, it will still take more than 2 cycles. To take a `u32add` operation as an example.

Plain `u32add` would take 4 cycles and would look like this:

```
u32assert2
u32add.unsafe
eq.0
assert
```

But `u32add.full` would take only 2 cycles and would look like this:
```
u32assert2
u32add.unsafe
```

PS: There is a typo in "we'd be able to do must safe operations at just 2x the cost of the unsafe ones" - "must" should be "most" :)

**Comment by 0xKanekiKen** [comment]
Thanks for the explanation. Makes sense now. What opcode shall I use for U32assert2?

**Comment by bobbinth** [comment]
For the assembly instruction let's call it `u32assert.2`.

As discussed above, as a part of the first PR there is no need to change the actual opcodes of the VM. But once we get there, the opcode could be just `u32assert2`.

**Comment by bobbinth** [comment]
@0xKanekiKen - yep, `u32assert2` sounds good.

**Comment by 0xKanekiKen** [comment]
@bobbinth - I've introduced `u32assert2` opcode to the VM instruction. Let me know your thought on the pull request.

**Comment by bobbinth** [comment]
Closed by #202
